web3.js/src/connection.ts

```diff
@@ -64,6 +66,133 @@ const BufferFromRawAccountData = coerce(
+  | {
+      state: 'subscribing';
+    }
+  // These subscriptions have been acknowledged by the
+  // server and have been assigned server subscription ids.
+  | {
+      serverSubscriptionId: ServerSubscriptionId;
+      state: 'subscribed';
+    }
+  // These subscriptions are intended to be torn down and
+  // are waiting on an acknowledgement from the server.
+  | {
+      serverSubscriptionId: ServerSubscriptionId;
+      state: 'unsubscribing';
+    }
+  // The request to tear down these subscriptions has been
+  // acknowledged by the server. The `serverSubscriptionId`
+  // is the id of the now-dead subscription.
+  | {
+      serverSubscriptionId: ServerSubscriptionId;
+      state: 'unsubscribed';
+    }
+>;
+/**
+ * Different RPC subscription methods accept different params.
+ */
+type SubscriptionConfig = Readonly<
+  | {
+      callback: AccountChangeCallback;
+      method: 'accountSubscribe';
+      params: Readonly<{
+        commitment?: Commitment; // Must default to 'finalized'
+        publicKey: string; // PublicKey of the account as a base 58 string
+      }>;
+      unsubscribeMethod: 'accountUnsubscribe';
+    }
+  | {
+      callback: LogsCallback;
+      method: 'logsSubscribe';
+      params: Readonly<{
+        commitment?: Commitment; // Must default to 'finalized'
+        filter: LogsFilter;
+      }>;
+      unsubscribeMethod: 'logsUnsubscribe';
+    }
+  | {
+      callback: ProgramAccountChangeCallback;
+      method: 'programSubscribe';
+      params: Readonly<{
+        commitment?: Commitment; // Must default to 'finalized'
+        filters?: GetProgramAccountsFilter[];
+        programId: string; // PublicKey of the program as a base 58 string
+      }>;
+      unsubscribeMethod: 'programUnsubscribe';
+    }
+  | {
+      callback: RootChangeCallback;
+      method: 'rootSubscribe';
+      params: undefined;
+      unsubscribeMethod: 'rootUnsubscribe';
+    }
+  | {
+      callback: SignatureSubscriptionCallback;
+      method: 'signatureSubscribe';
+      params: Readonly<{
+        signature: TransactionSignature; // TransactionSignature as a base 58 string
+        options?: SignatureSubscriptionOptions;
+      }>;
+      unsubscribeMethod: 'signatureUnsubscribe';
+    }
+  | {
+      callback: SlotChangeCallback;
+      method: 'slotSubscribe';
+      params: undefined;
+      unsubscribeMethod: 'slotUnsubscribe';
+    }
+  | {
+      callback: SlotUpdateCallback;
+      method: 'slotsUpdatesSubscribe';
+      params: undefined;
+      unsubscribeMethod: 'slotsUpdatesUnsubscribe';
+    }
+>;
+/**
+ * @internal
+ * Utility type that keeps tagged unions intact while omitting properties.
+ */
+type DistributiveOmit<T, K extends PropertyKey> = T extends unknown
+  ? Omit<T, K>
+  : never;
+type Subscription = BaseSubscription &
+  StatefulSubscription &
+  DistributiveOmit<SubscriptionConfig, 'callback'>;
+
+/**
+ * @internal
+ */
+function hashSubscriptionConfig({
+  method,
+  params,
```

Comment:
Can this operate over the "args" (what we pass to `this._rpcWebSocket.call(...)`) instead? That will more closely match the rpc server's behavior

Reply:
Nice catch. These all result in different server subscriptions, so best to push the hash function down, as you've said.

```
> await client.call('accountSubscribe',['CM78CPUeXjn8o3yroDHxUtKsZZgoy4GPkPPXfouKNH13', {encoding:'base64'}])
3
> await client.call('accountSubscribe',['CM78CPUeXjn8o3yroDHxUtKsZZgoy4GPkPPXfouKNH13', {encoding:'base64+zstd'}])
4
> await client.call('accountSubscribe',['CM78CPUeXjn8o3yroDHxUtKsZZgoy4GPkPPXfouKNH13', {encoding:'jsonParsed'}])
5
> await client.call('accountSubscribe',['CM78CPUeXjn8o3yroDHxUtKsZZgoy4GPkPPXfouKNH13', {encoding:'base58'}])
6
```